TESTS/mbed_hal/mpu/main.cpp

```diff
@@ -0,0 +1,175 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "greentea-client/test_env.h"
+
+#include "cmsis.h"
+#include <stdlib.h>
+
+#include "mpu_api.h"
+#include "mpu_test.h"
+
+#if !DEVICE_MPU
+#error [NOT_SUPPORTED] MPU API not supported for this target
+#endif
+
+using namespace utest::v1;
+
+#define HARDFAULT_IRQn   ((IRQn_Type)-13)
+
+// Assembly return instruction: bx lr
+#define ASM_BX_LR 0x4770
+
+volatile uint32_t fault_count;
+uint32_t real_hard_fault_handler;
+
+static volatile uint16_t data_function = ASM_BX_LR;
+static volatile uint16_t bss_function;
+
+static void call_mem(const volatile uint16_t *mem_function)
+{
+    // or the address with 1 to ensure the thumb bit is set
+    ((void (*)()) ((uint32_t)mem_function | 1))();
+}
+
+static void hard_fault_handler_test()
+{
+    fault_count++;
+    mbed_mpu_enable_ram_xn(false);
+}
+
+static void mpu_fault_test(const volatile uint16_t *mem_function)
+{
+    mbed_mpu_init();
+
+    // Verify that the mpu causes faults when executing ram
+    fault_count = 0;
+    mbed_mpu_enable_ram_xn(true);
+    call_mem(mem_function);
+    TEST_ASSERT_EQUAL(1, fault_count);
+
+    // Verify that the mpu can be turned off
+    fault_count = 0;
+    mbed_mpu_enable_ram_xn(false);
+    call_mem(mem_function);
+    TEST_ASSERT_EQUAL(0, fault_count);
+
+    // Verify that the mpu causes faults when executing ram
+    fault_count = 0;
+    mbed_mpu_enable_ram_xn(true);
+    call_mem(mem_function);
+    TEST_ASSERT_EQUAL(1, fault_count);
+
+    // Verify that free turns off the mpu
+    fault_count = 0;
+    mbed_mpu_free();
+    call_mem(mem_function);
+    TEST_ASSERT_EQUAL(0, fault_count);
+}
+
+void mpu_init_test()
+{
+    for (int i = 0; i < 10; i++) {
+        mbed_mpu_init();
+    }
+
+    mbed_mpu_free();
+}
+
+void mpu_free_test()
+{
+    mbed_mpu_init();
+
+    // Enable the MPU
+    mbed_mpu_enable_ram_xn(true);
+
+    // Free and ensure execution from RAM is allowed
+    mbed_mpu_free();
+
+    call_mem(&data_function);
+}
+
+void mpu_fault_test_data()
+{
+    mpu_fault_test(&data_function);
+}
+
+void mpu_fault_test_bss()
+{
+    bss_function = ASM_BX_LR;
+    mpu_fault_test(&bss_function);
+}
+
+void mpu_fault_test_stack()
+{
+    uint16_t stack_function;
+
+    stack_function = ASM_BX_LR;
+    mpu_fault_test(&stack_function);
+}
+
+void mpu_fault_test_heap()
+{
+    uint16_t *heap_function = (uint16_t*)malloc(2);
+
+    TEST_ASSERT_NOT_EQUAL(NULL, heap_function);
+    *heap_function = ASM_BX_LR;
+    mpu_fault_test(heap_function);
+
+    free(heap_function);
+}
+
+utest::v1::status_t fault_override_setup(const Case *const source, const size_t index_of_case)
+{
+    // Save old hard fault handler and replace it with a new one
+    real_hard_fault_handler = NVIC_GetVector(HARDFAULT_IRQn);
```

Comment:
Would it not be more direct to intercept MemManage exception? Would side-step the security issue.

(Any prospect of ever needing the combo of MPU + no ARMv8-M Main Extension? Not sure how easy it is to detect to conditionally support both)

Reply:
The `M2351` doesn't have `Main Extension` so the combo you mention is already a problem. I tried [enabling MemManage initially ](https://github.com/ARMmbed/mbed-os/pull/8633#issuecomment-435902779) but it didn't help as this device doesn't support it.

Reply:
I didn't see a "master" architecture flag for this. The [CMSIS MPU functions](https://github.com/ARMmbed/mbed-os/blob/master/cmsis/TARGET_CORTEX_M/mpu_armv8.h#L146) are checking for 
`SCB_SHCSR_MEMFAULTENA_Msk` being defined so I would probably use that if need be.